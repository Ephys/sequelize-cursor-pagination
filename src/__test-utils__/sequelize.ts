import type { Options } from '@sequelize/core';

export const TEST_databaseCredentials: Options = {
  dialect: 'postgres',
  port: 19_132,
  host: '0.0.0.0',
  password: 'password',
  username: 'user',
  database: 'db',
  logging: false,
};
