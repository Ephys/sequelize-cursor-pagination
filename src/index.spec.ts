import { DataTypes, ModelCtor, Op, Sequelize } from 'sequelize';
import { sequelizeFindByCursor } from '.';

let sequelize: Sequelize;
let userModel: ModelCtor<any>;

// TODO: ensure hasPreviousPage, hasNextPage is returning the correct value.

beforeAll(async () => {
  sequelize = new Sequelize({
    dialect: 'postgres',
    port: 19132,
    host: '0.0.0.0',
    password: 'password',
    username: 'user',
    database: 'db',
    logging: false,
  });

  userModel = sequelize.define('users', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    firstName: {
      field: 'first_name',
      type: DataTypes.TEXT,
      allowNull: false,
    },
    lastName: {
      field: 'last_name',
      type: DataTypes.TEXT,
      allowNull: false,
    },
    birthDate: {
      field: 'birth_date',
      type: DataTypes.DATEONLY,
      allowNull: false,
    },
  });

  await sequelize.sync({
    force: true,
  });

  await userModel.bulkCreate([{
    id: 5,
    firstName: 'Alan',
    lastName: 'LastName',
    birthDate: '2000-01-01',
  }, {
    id: 4,
    firstName: 'Bernard',
    lastName: 'LastName',
    birthDate: '1970-01-01',
  }, {
    id: 3,
    firstName: 'Cedric',
    lastName: 'Anderson',
    birthDate: '1980-01-01',
  }, {
    id: 2,
    firstName: 'Cedric',
    lastName: 'Brown',
    birthDate: '1960-01-01',
  }, {
    id: 6,
    firstName: 'Dimitri',
    lastName: 'LastName',
    birthDate: '1990-01-01',
  }, {
    id: 1,
    firstName: 'Dimitri',
    lastName: 'LastName',
    birthDate: '2010-01-01',
  }]);
});

describe('sequelizeFindByCursor', () => {
  it('includes PK in ORDER BY to ensures constant order', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 10,
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
      logging: query => {
        expect(query).toMatchSnapshot('query');
      },
    });

    // "Dimitri LastName" (1) should be before "Dimitri LastName" (6)

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(false);
  });

  it('does not include extra PK in ORDER BY if itâ€™s already present', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 10,
      order: [['firstName', 'ASC'], ['lastName', 'ASC'], ['id', 'DESC']],
      logging: query => {
        expect(query).toMatchSnapshot('query');
      },
    });

    // "Dimitri LastName" (6) should be before "Dimitri LastName" (1)

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(false);
  });

  it('supports returning the first x elements of the set', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 2,
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(true);
    expect(await results.hasPreviousPage()).toBe(false);
  });

  it('supports returning the last x elements of the set', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      last: 2,
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(true);
  });

  it('supports returning the first x elements after another cursor', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 2,
      after: {
        id: 3,
        firstName: 'Cedric',
        lastName: 'Anderson',
      },
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(true);
    expect(await results.hasPreviousPage()).toBe(true);
  });

  it('supports returning the last x elements after another cursor', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      last: 2,
      after: {
        id: 3,
        firstName: 'Cedric',
        lastName: 'Anderson',
      },
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(true);
  });

  it('supports returning the first x elements before another cursor', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 2,
      before: {
        id: 3,
        firstName: 'Cedric',
        lastName: 'Anderson',
      },
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(false);
  });

  it('supports returning the last x elements before another cursor', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      last: 2,
      before: {
        id: 1,
        firstName: 'Dimitri',
        lastName: 'LastName',
      },
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(true);
    expect(await results.hasPreviousPage()).toBe(true);
  });

  it('supports filtering', async () => {
    const results = await sequelizeFindByCursor({
      attributes: ['firstName', 'lastName', 'id'],
      model: userModel,
      first: 10,
      after: {
        id: 4,
        firstName: 'Bernard',
        lastName: 'LastName',
      },
      // this WHERE will be joined to the cursor filter using AND
      where: {
        birthDate: { [Op.gte]: '1990-01-01' },
      },
      order: [['firstName', 'ASC'], ['lastName', 'ASC']],
      logging: sql => {
        expect(sql).toMatchSnapshot('query');
      },
    });

    expect(results.nodes).toMatchSnapshot('nodes');
    expect(await results.hasNextPage()).toBe(false);
    expect(await results.hasPreviousPage()).toBe(true);
  });
});

afterAll(() => {
  return sequelize.close();
});
