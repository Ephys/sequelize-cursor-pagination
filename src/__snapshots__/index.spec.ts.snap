// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sequelizeFindByCursor does not include extra PK in ORDER BY if it’s already present: nodes 1`] = `
[
  {
    "firstName": "Alan",
    "id": 5,
    "lastName": "LastName",
  },
  {
    "firstName": "Bernard",
    "id": 4,
    "lastName": "LastName",
  },
  {
    "firstName": "Cedric",
    "id": 3,
    "lastName": "Anderson",
  },
  {
    "firstName": "Cedric",
    "id": 2,
    "lastName": "Brown",
  },
  {
    "firstName": "Dimitri",
    "id": 6,
    "lastName": "LastName",
  },
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor does not include extra PK in ORDER BY if it’s already present: query 1`] = `"Executing (default): SELECT "users".* FROM (SELECT "first_name" AS "firstName", "last_name" AS "lastName", "id" FROM "users" AS "users" ORDER BY "users"."first_name" ASC, "users"."last_name" ASC, "users"."id" DESC LIMIT 11) AS "users" ORDER BY "firstName" ASC, "lastName" ASC, "users"."id" DESC;"`;

exports[`sequelizeFindByCursor includes PK in ORDER BY to ensures constant order: nodes 1`] = `
[
  {
    "firstName": "Alan",
    "id": 5,
    "lastName": "LastName",
  },
  {
    "firstName": "Bernard",
    "id": 4,
    "lastName": "LastName",
  },
  {
    "firstName": "Cedric",
    "id": 3,
    "lastName": "Anderson",
  },
  {
    "firstName": "Cedric",
    "id": 2,
    "lastName": "Brown",
  },
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
  {
    "firstName": "Dimitri",
    "id": 6,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor includes PK in ORDER BY to ensures constant order: query 1`] = `"Executing (default): SELECT "users".* FROM (SELECT "first_name" AS "firstName", "last_name" AS "lastName", "id" FROM "users" AS "users" ORDER BY "users"."first_name" ASC, "users"."last_name" ASC, "users"."id" ASC LIMIT 11) AS "users" ORDER BY "firstName" ASC, "lastName" ASC, "users"."id" ASC;"`;

exports[`sequelizeFindByCursor supports filtering: nodes 1`] = `
[
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
  {
    "firstName": "Dimitri",
    "id": 6,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor supports filtering: query 1`] = `"Executing (default): SELECT "users".* FROM (SELECT "first_name" AS "firstName", "last_name" AS "lastName", "id" FROM "users" AS "users" WHERE ("users"."birth_date" >= '1990-01-01' AND ("users"."first_name" > 'Bernard' OR ("users"."first_name" = 'Bernard' AND ("users"."last_name" > 'LastName' OR ("users"."last_name" = 'LastName' AND "users"."id" > 4))))) ORDER BY "users"."first_name" ASC, "users"."last_name" ASC, "users"."id" ASC LIMIT 11) AS "users" ORDER BY "firstName" ASC, "lastName" ASC, "users"."id" ASC;"`;

exports[`sequelizeFindByCursor supports filtering: query 2`] = `"Executing (default): SELECT "users".* FROM (SELECT "first_name" AS "firstName", "last_name" AS "lastName", "id" FROM "users" AS "users" WHERE ("users"."birth_date" >= '1990-01-01' AND ("users"."first_name" < 'Bernard' OR ("users"."first_name" = 'Bernard' AND ("users"."last_name" < 'LastName' OR ("users"."last_name" = 'LastName' AND "users"."id" < 4))))) ORDER BY "users"."first_name" DESC, "users"."last_name" DESC, "users"."id" DESC LIMIT 1) AS "users" ORDER BY "firstName" DESC, "lastName" DESC, "users"."id" DESC;"`;

exports[`sequelizeFindByCursor supports returning the first x elements after another cursor: nodes 1`] = `
[
  {
    "firstName": "Cedric",
    "id": 2,
    "lastName": "Brown",
  },
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor supports returning the first x elements before another cursor: nodes 1`] = `
[
  {
    "firstName": "Alan",
    "id": 5,
    "lastName": "LastName",
  },
  {
    "firstName": "Bernard",
    "id": 4,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor supports returning the first x elements of the set: nodes 1`] = `
[
  {
    "firstName": "Alan",
    "id": 5,
    "lastName": "LastName",
  },
  {
    "firstName": "Bernard",
    "id": 4,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor supports returning the last x elements after another cursor: nodes 1`] = `
[
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
  {
    "firstName": "Dimitri",
    "id": 6,
    "lastName": "LastName",
  },
]
`;

exports[`sequelizeFindByCursor supports returning the last x elements before another cursor: nodes 1`] = `
[
  {
    "firstName": "Cedric",
    "id": 3,
    "lastName": "Anderson",
  },
  {
    "firstName": "Cedric",
    "id": 2,
    "lastName": "Brown",
  },
]
`;

exports[`sequelizeFindByCursor supports returning the last x elements of the set: nodes 1`] = `
[
  {
    "firstName": "Dimitri",
    "id": 1,
    "lastName": "LastName",
  },
  {
    "firstName": "Dimitri",
    "id": 6,
    "lastName": "LastName",
  },
]
`;
